src/annepro2.rs

```diff
@@ -57,57 +57,75 @@ pub fn flash_firmware<R: std::io::Read>(
     file: &mut R,
     boot: bool,
 ) -> std::result::Result<(), AP2FlashError> {
-    match HidApi::new() {
-        Ok(api) => {
-            for dev in api.device_list() {
-                println!(
-                    "HID Dev: {:04x}:{:04x} usage #: {:02x} usage_page #: {:04x} {:?}",
-                    dev.vendor_id(),
-                    dev.product_id(),
-                    dev.usage(),
-                    dev.usage_page(),
-                    dev.product_string()
-                );
-            }
-            let dev = api
-                .device_list()
-                .find(|dev| {
-                    dev.vendor_id() == ANNEPRO2_VID
-                        && ((dev.product_id() == PID_C15 && dev.interface_number() == 1)
-                            || (dev.product_id() == PID_C18))
-                })
-                .expect("No device found. Please remind to put the device into IAP Mode");
-
-            let handle = dev.open_device(&api).expect("unable to open device");
-            handle.set_blocking_mode(true).expect("non-blocking");
-            println!(
-                "device is {:?}",
-                handle.get_product_string().expect("string")
-            );
-
-            // Flashing Code
-            erase_device(&handle, target, base).map_err(|err| {
-                println!("Error while erasing: {}", err);
-                AP2FlashError::USBError
-            })?;
-            flash_file(&handle, target, base, file);
-            write_ap_flag(&handle, 2).map_err(|e| {
-                println!("Error while writing AP flag: {:?}", e);
-                AP2FlashError::USBError
-            })?;
-            if boot {
-                boot_device(&handle).map_err(|e| {
-                    println!("Error while booting device: {:?}", e);
-                    AP2FlashError::USBError
-                })?;
-            }
-            Ok(())
-        }
-        Err(e) => {
-            println!("Error: {:?}", e);
-            Err(AP2FlashError::USBError)
+    let api = HidApi::new().map_err(|_| AP2FlashError::USBError)?;
+
+    let (anne_devices, flash_device) = fetch_devices(&api);
```

Comment:
technically, instead of returning all the `anne_devices` here, I could simply return a Boolean like `has_anne_devices`, if it makes cleaner and clearer

Reply:
Also, instead of returning a tuple, if we want named fields I wouldn't mind creating a `struct` specifically for the result, if we don't think it would pollute of a semi-unnecessary type